tools/tgstation-server-3/TGServerService/Interop.cs

```diff
@@ -0,0 +1,202 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Net;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading;
+using System.Web.Security;
+using TGServiceInterface;
+
+namespace TGServerService
+{
+	//handles talking between the world and us
+	partial class TGStationServer
+	{
+		QueuedLock topicLock = new QueuedLock();
+		Socket topicSender = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp) { SendTimeout = 5000, ReceiveTimeout = 5000 };
+
+		const int CommsKeyLen = 64;
+		string serviceCommsKey; //regenerated every DD restart
+
+		Thread NudgeThread;
+		object NudgeLock = new object();
+
+		//See code/modules/server_tools/server_tools.dm for command switch
+		const string SCHardReboot = "hard_reboot";  //requests that dreamdaemon restarts when the round ends
+		const string SCGracefulShutdown = "graceful_shutdown";  //requests that dreamdaemon stops when the round ends
+		const string SCWorldAnnounce = "world_announce";	//sends param 'message' to the world
+		const string SCIRCCheck = "irc_check";  //returns game stats
+		const string SCIRCStatus = "irc_status";	//returns admin stats
+		const string SCNameCheck = "namecheck";	//returns keywords lookup
+		const string SCAdminPM = "adminmsg";	//pms a target ckey
+		const string SCAdminWho = "adminwho";	//lists admins
+
+		//raw command string sent here via world.ExportService
+		void HandleCommand(string cmd)
+		{
+			var splits = new List<string>(cmd.Split(' '));
+
+			switch (splits[0])
+			{
+				case "irc":
+					splits.RemoveAt(0);
+					SendMessage("GAME: " + String.Join(" ", splits));
+					break;
+				case "killme":
+					Restart();
+					break;
+				case "send2irc":
+					splits.RemoveAt(0);
+					SendMessage("RELAY: " + String.Join(" ", splits), true);
+					break;
+			}
+		}
+		
+		string SendCommand(string cmd)
+		{
+			lock (watchdogLock)
+			{
+				if (currentStatus != TGDreamDaemonStatus.Online)
+					return "Error: Server Offline!";
+				return SendTopic(String.Format("serviceCommsKey={0};command={1}", serviceCommsKey, cmd), currentPort);
+			}
+		}
+
+		string SendPM(string targetCkey, string sender, string message)
+		{
+			return SendCommand(String.Format("{3};target={0};sender={1};message={2}", targetCkey, sender, message, SCAdminPM));
+		}
+
+		string NameCheck(string targetCkey, string sender)
+		{
+			return SendCommand(String.Format("{2};target={0};sender={1}", targetCkey, sender, SCNameCheck));
+		}
+
+		//Fuckery to diddle byond with the right packet to accept our girth
+		string SendTopic(string topicdata, ushort port, bool retry = false)

```

Comment:
The byond packet structure is as follows:

2 bytes sequence id(only after first packet, not used here), 2 bytes command id, 2 bytes content length, content.

Your system only encodes the second length byte, so you would be limited to 255 characters. i think most versions of this system has this same flaw, but otherwise it should be fine.

The 5 empty bytes at the front are flags and id'ing the world that sent the topic, not used in our system (there is also a way to send a file using this, but I don't understand it, and there is an encrypted version that has persistent connection support, but goonstation is keeping the documentation of that that they figured out secret so they can use it to lock support of goonhub into their control.).

This is a complete (and slightly more proper) packet in php:

```php
$packet = "\x00\x83" . pack('n', strlen($str) + 6) . "\x00\x00\x00\x00\x00" . $str . "\x00";
```